### ISSUE TITLE
Traefik won't add auto restarted container to backend

### ISSUE THREAD

**Issue by minhdanh** [containous/traefik#3780]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

We have Traefik and a Drone server running in Docker (with Docker Compose).
After upgrading Traefik to version 1.7 (image: `traefik:v1.7`), sometimes when the Drone server exits because of a mysterious bug, Drone is restarted automatically by Docker Compose. But Traefik would fail to add Drone to backend again.
At this time we have to restart Treafik manually.

### What did you expect to see?
Traefik would see restarted container and add it to backend correctly.

Traefik logs:
```
time="2018-07-27T11:01:56Z" level=warning msg="Unable to find the IP address for the container \"/drone-server\": the server is ignored."
time="2018-07-27T11:01:56Z" level=error msg="undefined backend 'backend-drone-server' for frontend frontend-Host-ci-example-com-0. Skipping frontend frontend-Host-ci-example
```

### What did you see instead?
Traefik failed to get the Drone container's IP, and won't add it to backend.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
v1.7.0-rc2
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
logLevel = "INFO"

defaultEntryPoints = ["http", "https"]

checkNewVersion = false
InsecureSkipVerify = true

[metrics]
  [metrics.prometheus]
    entryPoint = "traefik"
    buckets = [0.1,0.3,1.2,5.0]

[api]
  entryPoint = "traefik"
  dashboard = true

[docker]
exposedbydefault = false

[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
    entryPoint = "https"
  [entryPoints.https]
  address = ":443"
    [entryPoints.https.tls]
      [[entryPoints.https.tls.certificates]]
      certFile = "/etc/traefik/tls.cert"
      keyFile = "/etc/traefik/tls.key"

[acme]
storage = "/data/acme.json"
email = "email@example.com"
entryPoint = "https"
onHostRule = true
ondemand = false
  [acme.httpChallenge]
  entryPoint = "http"
```

docker-compose.yml:
```yml
version: '3.4'

  drone-server:
    container_name: drone-server
    image: drone/drone:0.8.5
    restart: unless-stopped
    ports:
      - "9000:9000"

﻿  proxy:
    container_name: proxy
    # https://hub.docker.com/r/library/traefik/tags/
    image: traefik:v1.7
    restart: unless-stopped
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./traefik.toml:/etc/traefik/traefik.toml
```

**Comment by dduportal** [comment]
Hello @minhdanh, thanks for your interest in the project.

We are missing information to reproduce this issue. I was able to setup a working stack with this configuration: https://gist.github.com/dduportal/e744ebe0cff33c0a1d590f572b6a29f1 , on my Docker4Mac 18.06 and on an Ubuntu 16.04 with Docker 18.03. 
Both setups were working as expected when sending a sigkill signal to the drone-process, which restarted the container. I've ensured that the container private IP changed before checking the behavior: in each case, traefik was picking up the change correctly.

It looks like you are missing some labels (which you can check on provided gist) which might be important:

- https://gist.github.com/dduportal/e744ebe0cff33c0a1d590f572b6a29f1#file-docker-compose-yml-L14 (`traefik.enable`) since you disabled the "defaultExpose" on your traefik.toml file.
- https://gist.github.com/dduportal/e744ebe0cff33c0a1d590f572b6a29f1#file-docker-compose-yml-L17 (`traefik.port`) because Drone's container is exposing different ports by default (ref. https://github.com/drone/drone/blob/master/Dockerfile#L4) so we need to help Traefik to guess which port to use for the backend with this label

Can you confirm that even with the provided examples for labels, you still have the issue? If yes, we need more informations:

* Docker engine used, which OS, configuration of the Docker engine
* the whole docker-compose.yml configuration you used
* the whole traefik.log

Thanks in advance!

**Comment by thomasf** [comment]
I see similar problems after upgrading to 1.7.2 from 1.6..`docker-compose up -d` often makes the application routing disappear. `docker-compose restart someapp someotherapp` usually brings them up again.. It has happened on 4-5 hosts several times, I will try to see if I can replicate it under suitable logging conditions..

**Comment by thomasf** [comment]
Should I create a new issue or use this one?

**Comment by huafu** [comment]
I can confirm this issue, not sure it's exactly the same, but as @thomasf it's randomly not restored. It's often, but not always, when I do a `docker-compose pull && docker-compose --no-build up` after pushing new images from my local machine.

Sounds like the down/up event are too fast for traefik to handle them both (but IDK how it works under the hood, just guessing here)

**Comment by jortkoopmans** [comment]
I've been struggling with this for a while, at least...I think it's similar to what is reported in this issue by several of you. The failure behavior is intermittent.
If for any reason a container is restarted, there is a chance that the traefik routes will not be re-established.

I can reproduce this intermittent behavior by simply setting up traefik and an nginx container, and then restarting the nginx container a couple of times (docker restart someapp).
I'm suspecting a timing issue.

A successful, normal (working) restart, shows in Traefik as;

> time="2018-10-19T13:12:09Z" level=debug msg="Provider event received {Status:die ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 From:nginx Type:container Action:die Actor:{ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 Attributes:map[maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.frontend.entryPoints:http,https com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.oneoff:False com.docker.compose.version:1.22.0 traefik.frontend.rule:HostRegexp:{catchall:.*} com.docker.compose.project:temp image:nginx name:someapp com.docker.compose.container-number:1 com.docker.compose.service:someapp exitCode:0 traefik.backend:someapp traefik.port:80]} Scope:local Time:1539954729 TimeNano:1539954729940309118}"
time="2018-10-19T13:12:10Z" level=debug msg="Filtering disabled container /traefik"
time="2018-10-19T13:12:10Z" level=debug msg="Configuration received from provider docker: {}"
time="2018-10-19T13:12:10Z" level=debug msg="Creating entry point redirect http -> https"
time="2018-10-19T13:12:10Z" level=info msg="Server configuration reloaded on :8080"
time="2018-10-19T13:12:10Z" level=info msg="Server configuration reloaded on :80"
time="2018-10-19T13:12:10Z" level=info msg="Server configuration reloaded on :443"
time="2018-10-19T13:12:10Z" level=debug msg="Provider event received {Status:start ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 From:nginx Type:container Action:start Actor:{ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 Attributes:map[image:nginx name:someapp traefik.backend:someapp traefik.frontend.entryPoints:http,https com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.project:temp com.docker.compose.version:1.22.0 maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.port:80 com.docker.compose.container-number:1 com.docker.compose.oneoff:False com.docker.compose.service:someapp]} Scope:local Time:1539954730 TimeNano:1539954730278104338}"
time="2018-10-19T13:12:10Z" level=debug msg="Filtering disabled container /traefik"
time="2018-10-19T13:12:10Z" level=debug msg="originLabelsmap[com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.version:1.22.0 maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.frontend.rule:HostRegexp:{catchall:.*} com.docker.compose.container-number:1 com.docker.compose.oneoff:False com.docker.compose.project:temp com.docker.compose.service:someapp traefik.backend:someapp traefik.frontend.entryPoints:http,https traefik.port:80]"
time="2018-10-19T13:12:10Z" level=debug msg="allLabelsmap[:map[traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.backend:someapp traefik.frontend.entryPoints:http,https traefik.port:80]]"
time="2018-10-19T13:12:10Z" level=debug msg="originLabelsmap[com.docker.compose.container-number:1 com.docker.compose.oneoff:False com.docker.compose.project:temp com.docker.compose.service:someapp traefik.backend:someapp traefik.frontend.entryPoints:http,https traefik.port:80 com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.version:1.22.0 maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.frontend.rule:HostRegexp:{catchall:.*}]"
time="2018-10-19T13:12:10Z" level=debug msg="allLabelsmap[:map[traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.port:80 traefik.backend:someapp traefik.frontend.entryPoints:http,https]]"
time="2018-10-19T13:12:10Z" level=debug msg="Backend backend-someapp: no load-balancer defined, fallback to 'wrr' method"
time="2018-10-19T13:12:10Z" level=debug msg="Configuration received from provider docker: {\"backends\":{\"backend-someapp\":{\"servers\":{\"server-someapp-9121159f5bca979e8087544fd72045a9\":{\"url\":\"http://192.168.32.3:80\",\"weight\":1}},\"loadBalancer\":{\"method\":\"wrr\"}}},\"frontends\":{\"frontend-HostRegexp-catchall-0\":{\"entryPoints\":[\"http\",\"https\"],\"backend\":\"backend-someapp\",\"routes\":{\"route-frontend-HostRegexp-catchall-0\":{\"rule\":\"HostRegexp:{catchall:.*}\"}},\"passHostHeader\":true,\"priority\":0,\"basicAuth\":null}}}"
time="2018-10-19T13:12:12Z" level=debug msg="Creating entry point redirect http -> https"
time="2018-10-19T13:12:12Z" level=debug msg="Wiring frontend frontend-HostRegexp-catchall-0 to entryPoint http"
time="2018-10-19T13:12:12Z" level=debug msg="Creating backend backend-someapp"
time="2018-10-19T13:12:12Z" level=debug msg="Adding TLSClientHeaders middleware for frontend frontend-HostRegexp-catchall-0"
time="2018-10-19T13:12:12Z" level=debug msg="Creating load-balancer wrr"
time="2018-10-19T13:12:12Z" level=debug msg="Creating server server-someapp-9121159f5bca979e8087544fd72045a9 at http://192.168.32.3:80 with weight 1"
time="2018-10-19T13:12:12Z" level=debug msg="Creating route route-frontend-HostRegexp-catchall-0 HostRegexp:{catchall:.*}"
time="2018-10-19T13:12:12Z" level=debug msg="Wiring frontend frontend-HostRegexp-catchall-0 to entryPoint https"
time="2018-10-19T13:12:12Z" level=debug msg="Creating backend backend-someapp"
time="2018-10-19T13:12:12Z" level=debug msg="Adding TLSClientHeaders middleware for frontend frontend-HostRegexp-catchall-0"
time="2018-10-19T13:12:12Z" level=debug msg="Creating load-balancer wrr"
time="2018-10-19T13:12:12Z" level=debug msg="Creating server server-someapp-9121159f5bca979e8087544fd72045a9 at http://192.168.32.3:80 with weight 1"
time="2018-10-19T13:12:12Z" level=debug msg="Creating route route-frontend-HostRegexp-catchall-0 HostRegexp:{catchall:.*}"
time="2018-10-19T13:12:12Z" level=info msg="Server configuration reloaded on :8080"
time="2018-10-19T13:12:12Z" level=info msg="Server configuration reloaded on :80"
time="2018-10-19T13:12:12Z" level=info msg="Server configuration reloaded on :443"


While a unsuccessful restart (broken) shows:

>time="2018-10-19T13:07:57Z" level=debug msg="Provider event received {Status:die ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 From:nginx Type:container Action:die Actor:{ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 Attributes:map[name:someapp traefik.frontend.rule:HostRegexp:{catchall:.*} com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.project:temp maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.backend:someapp traefik.port:80 com.docker.compose.container-number:1 com.docker.compose.version:1.22.0 exitCode:0 traefik.frontend.entryPoints:http,https com.docker.compose.oneoff:False com.docker.compose.service:someapp image:nginx]} Scope:local Time:1539954477 TimeNano:1539954477012610712}"
time="2018-10-19T13:07:57Z" level=debug msg="Filtering disabled container /traefik"
time="2018-10-19T13:07:57Z" level=debug msg="Configuration received from provider docker: {}"
time="2018-10-19T13:07:57Z" level=debug msg="Creating entry point redirect http -> https"
time="2018-10-19T13:07:57Z" level=debug msg="Provider event received {Status:start ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 From:nginx Type:container Action:start Actor:{ID:f3fde9c2dba7db19a0f93ddeb561886bf256be66dc5577543618007ae3618150 Attributes:map[com.docker.compose.container-number:1 com.docker.compose.service:someapp maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> name:someapp traefik.backend:someapp traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.port:80 com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.oneoff:False com.docker.compose.project:temp com.docker.compose.version:1.22.0 image:nginx traefik.frontend.entryPoints:http,https]} Scope:local Time:1539954477 TimeNano:1539954477343667521}"
time="2018-10-19T13:07:57Z" level=debug msg="Filtering disabled container /traefik"
time="2018-10-19T13:07:57Z" level=debug msg="originLabelsmap[com.docker.compose.version:1.22.0 maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.backend:someapp com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.container-number:1 com.docker.compose.oneoff:False com.docker.compose.project:temp com.docker.compose.service:someapp traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.frontend.entryPoints:http,https traefik.port:80]"
time="2018-10-19T13:07:57Z" level=debug msg="allLabelsmap[:map[traefik.frontend.entryPoints:http,https traefik.port:80 traefik.backend:someapp traefik.frontend.rule:HostRegexp:{catchall:.*}]]"
time="2018-10-19T13:07:57Z" level=debug msg="originLabelsmap[com.docker.compose.config-hash:b2b6bad273c2aab4e14fef131d833e586f292e4a56b35863fc0b12775ce5b53a com.docker.compose.container-number:1 com.docker.compose.oneoff:False com.docker.compose.project:temp com.docker.compose.service:someapp com.docker.compose.version:1.22.0 maintainer:NGINX Docker Maintainers <docker-maint@nginx.com> traefik.backend:someapp traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.frontend.entryPoints:http,https traefik.port:80]"
time="2018-10-19T13:07:57Z" level=debug msg="allLabelsmap[:map[traefik.backend:someapp traefik.frontend.rule:HostRegexp:{catchall:.*} traefik.frontend.entryPoints:http,https traefik.port:80]]"
time="2018-10-19T13:07:57Z" level=debug msg="Backend backend-someapp: no load-balancer defined, fallback to 'wrr' method"
time="2018-10-19T13:07:57Z" level=debug msg="Configuration received from provider docker: {\"backends\":{\"backend-someapp\":{\"servers\":{\"server-someapp-9121159f5bca979e8087544fd72045a9\":{\"url\":\"http://192.168.32.3:80\",\"weight\":1}},\"loadBalancer\":{\"method\":\"wrr\"}}},\"frontends\":{\"frontend-HostRegexp-catchall-0\":{\"entryPoints\":[\"http\",\"https\"],\"backend\":\"backend-someapp\",\"routes\":{\"route-frontend-HostRegexp-catchall-0\":{\"rule\":\"HostRegexp:{catchall:.*}\"}},\"passHostHeader\":true,\"priority\":0,\"basicAuth\":null}}}"
time="2018-10-19T13:07:57Z" level=info msg="Skipping same configuration for provider docker"
time="2018-10-19T13:07:57Z" level=info msg="Server configuration reloaded on :80"
time="2018-10-19T13:07:57Z" level=info msg="Server configuration reloaded on :443"
time="2018-10-19T13:07:57Z" level=info msg="Server configuration reloaded on :8080"

To me, it looks like the missing 'reload' of the Server configuration, after the container died, has something to do with it.

To facilitate further troubleshooting, I've attached various files (in zip):
working.log (see also above)
broken.log (see also above)
docker-compose.yml
traefik.toml
docker-info.log
[traefik_3780.zip](https://github.com/containous/traefik/files/2496037/traefik_3780.zip)


uname -a:
Linux HOSTNAME 3.10.0-862.3.2.el7.x86_64 #1 SMP Mon May 21 23:36:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
cat /etc/redhat-release
CentOS Linux release 7.5.1804 (Core)

Traefik image 'latest'
Digest sha256:c25ac3c2736dc41b08c6f40ef3393d03dd7640a037cd53d681af459ccfa95c28

**Comment by thomasf** [comment]
I have a feeliung that this is mostly happening on hosts with less capability but I don't have any hard data to back this up.. We run a bunch of the two lowest tier linodes (between 1-2 cores and 1-2GB of ram) and it seems to happen on those a lot more often than on mor capable hosts with 8+ cores and lots of ram.. Spontaneously this feels like some kind of race condition somewhere...

**Comment by dduportal** [comment]
Thanks @jortkoopmans , we are currently trying to reproduce the issue.

Based on your feedback, I tried on Docker4Mac (with 1 CPU 1 Gb) with similar docker-compose.yml, but I did not see anything as you described. However the disk is an SSD, so I'm gonna try on slower machines on Digital Ocean and AWS to reproduce.

**Comment by chrootlogin** [comment]
I have this problem since a while. It's really annoying and it was not there in the beginning. What's interesting is, that when I run https://github.com/spotify/docker-gc after the update of the container, it starts working again.

**Comment by chrootlogin** [comment]
I found out that the problem was introduced in v1.7. When I use the old image with v1.6 it works again, but with v1.7 it comes back.

**Comment by jortkoopmans** [comment]
Certainly, please see the daemon (with debug on) logs with concatenated traefik logs, attached. The two files represent both a successful and failing restart of 'someapp' (nginx container).
I've updated docker to 18.09 (ce), running the latest traefik container v1.7.4 for this test. I used the same docker-compose config as posted earlier.

To me it seems the only difference is the timing/ordering of several calls.
I hope this helps to pinpoint the issue.

[logs_failing.txt](https://github.com/containous/traefik/files/2561304/logs_failing.txt)
[logs_working.txt](https://github.com/containous/traefik/files/2561305/logs_working.txt)

**Comment by jortkoopmans** [comment]
I can also reproduce this problem using traefik v1.7.7 (on CentOS 3.10.0-957).

**Comment by jortkoopmans** [comment]
@minhdanh @thomasf ; Which OS and kernel did you use?
@dduportal ; Did you have a chance to reproduce this issue using CentOS?

**Comment by daniellavoie** [comment]
I also encountered the : `Unable to find the IP address for the container ... : the server is ignored.` error message.

After some troubleshooting I realized that my container was misbehaving and was continuously restarting. As such traefik did not have the time to set up a backend for the container.

**Comment by gregoryelmajjouti** [comment]
We have also encountered this issue (as reported by @jortkoopmans) on the 1.6.6 release of Traefik when running on CentOS. It does not seem to be a regression in 1.7.

I think we have isolated the race condition to the following function: 
https://github.com/containous/traefik/blob/246b245959f1a16d78a17f13e9bc659bcafb3d47/server/server_configuration.go#L149-L179

If a Docker container is restarted, it sends a `die` and a `start` provider event. These events are turned into `config.Message` types and they get processed by this function in quick succession. The race is between the read at line 164 and the write at line 178 which seem to be running in parallel for both config messages.

The race condition happens if during the processing of the `start` event the `deepEqual` at line 164 sees the current configuration state before the remove triggered by the `die` message is applied (line 178). The result is that return on line 166 drops the `start` configuration change. This means only the `die` message is being processed, and the Docker container no longer exists according to Traefik.

A simple workaround that works for us is to comment out the return on line 166, which means the `start` is processed. The throttle in `providerConfigUpdateCh` (applied on line 174) already prevents too many configuration changes.

**Comment by jortkoopmans** [comment]
I can still reproduce this using;
Docker CE 18.09.6
Traefik v1.7.11

**Comment by olivierHa** [comment]
Same here, I got this issue using "docker-compose" with Traefik v1.7.11 and docker 18.06.1-ce.

**Comment by dtomcej** [comment]
@jortkoopmans, does the fix provided by @gregoryelmajjouti resolve your issue?

If so, we probably want to have a PR to mainline this fix, if it is still an issue...

**Comment by wader** [comment]
Sam here. Seen this issue several times with 1.7.10. Restarting the container again usually fixes it.

**Comment by showcheap** [comment]
Same issue with Amazon Linux (centos) on AWS ec2 1GB Ram & SSD. Traefik version 1.7.12 still get same issue when updating images (docker build & docker-compose up -d <servicename>). After updating to 1.7.14  and try to update container few time, issue disapear, I hope 1.7.14 fix this issue.

**Comment by jortkoopmans** [comment]
> @jortkoopmans, does the fix provided by @gregoryelmajjouti resolve your issue?
> 
> If so, we probably want to have a PR to mainline this fix, if it is still an issue...

Sorry for delay, I have finally got around to build traefik after disabling the 'return' line (I'm using up to date 1.7 branch), as suggested by @gregoryelmajjouti. I can confirm this **solves the problem**. 
However the unit-tests are failing;
```
--- FAIL: TestListenProvidersSkipsSameConfigurationForProvider (0.12s)
    server_test.go:137: Same configuration should not be published multiple times
```

I'm not a (go) developer, so I won't be able to provide a PR which also addresses the unit tests.



> After updating to 1.7.14 and try to update container few time, issue disapear, I hope 1.7.14 fix this issue.

I have tested on 1.7.14, the problem is unfortunately still there. Since its a race condition, you may not have hit it yet (i.e. run some more tests to break it).

**Comment by showcheap** [comment]
@jortkoopmans that's right, issue still exist in 1.7.14 after several tests.

**Comment by dduportal** [comment]
Hi @jortkoopmans @showcheap , sorry for this inconvenience, and thanks for your feedbacks.
Have you tried to reproduce with the `v2.0.0-rc1` version of Traefik?

We understand that it requires to adapt the config (and migration guides are currently worked on), but with <https://docs.traefik.io/v2.0/providers/docker/> you might be able to test if the issue is still there in 2.0. Let us know :)

**Comment by showcheap** [comment]
@dduportal trying on `traefik:2.0` docker images today, and the issue still exists. Maybe I need to switch my OS for now.

**Comment by dduportal** [comment]
@showcheap what is the Docker CE Engine version you're using?

**Comment by showcheap** [comment]
@dduportal 

```
yum info docker
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
Installed Packages
Name        : docker
Arch        : x86_64
Version     : 18.06.1ce
Release     : 10.amzn2
Size        : 151 M
Repo        : installed
From repo   : amzn2extra-docker
Summary     : Automates deployment of containerized applications
URL         : http://www.docker.com
License     : ASL 2.0 and MIT and BSD and MPLv2.0 and WTFPL
Description : Docker is an open-source engine that automates the deployment of any
            : application as a lightweight, portable, self-sufficient container that will
            : run virtually anywhere.
            : 
            : Docker containers can encapsulate any payload, and will run consistently on
            : and between virtually any server. The same container that a developer builds
            : and tests on a laptop will run at scale, in production*, on VMs, bare-metal
            : servers, OpenStack clusters, public instances, or combinations of the above.
```

**Comment by jortkoopmans** [comment]
Ok @dduportal , I've also gone ahead and tried with v2.0 branch as well as the readily available traefik:v2.0 image; The **same problem occurs there**.

Then I removed the 'return' line, similar to what was done for the 1.7 testing (but now the file lives here https://github.com/containous/traefik/blob/v2.0/pkg/server/server_configuration.go#L216) and rebuilt it. The code does not seem touched here, removing the line similarly breaks the unit tests as on 1.7.
When the 'return' is removed, the problem is resolved (i.e. the container is always properly restarted).

I don't think the docker version will really matter, given the race condition is really within the application, but for reference;

```
$docker info
Client:
 Debug Mode: false

Server:
 Containers: 6
  Running: 4
  Paused: 0
  Stopped: 2
 Images: 45
 Server Version: 19.03.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-957.12.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.638GiB
 Name: somehostname
 ID: FDRQ:4H64:GYEH:MZ3K:ULW6:HZNK:SIRS:OGTD:KBLR:JLK4:IC4I:FMUP
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 58
  Goroutines: 62
  System Time: 2019-08-29T13:29:44.741524137Z
  EventsListeners: 1
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Comment by dduportal** [comment]
Thanks for your feedbacks!

The questions about OS and Docker versions are aimed at providing as much context as possible to reproduce the issue, which is the requirement to fix it in the future.
Same for the `2.0` version, as the codebase changed quite a lot: it helps to gather conditions.

@jortkoopmans @gregoryelmajjouti thanks for the deep technical elements. The developer team will look into this, but no ETA given that it's hard for us to reproduce and analyse. As the proposed workaround is breaking tests, it means that it might solve this issue, but could lead to greater trouble somewhere else, hence the required analysis.

**Comment by jortkoopmans** [comment]
Just thought about running the tests again on the now officially released Traefik v2.0.1 --> I can confirm the problem is still occurring.

For convenience of reproducing this issue, please see the files attached (traefik v2 syntax).
It is a docker-compose file, to bring up traefik and a containous/whoami container. Using the .env for some simple settings (you likely don't need to change anything there, but you can) and a small script to kick the container and verify successful or broken addition to the Traefik routes.
Should be as simple as a 'docker-compose up -d' and running the kick script. I usually break in the first 10 cycles, but your mileage may vary.

[traefik_v2_issue3780.zip](https://github.com/containous/traefik/files/3689494/traefik_v2_issue3780.zip)

**Comment by thomasf** [comment]
I just had it happen with latest 1.7 version in production a few times yesterday.

I can also confirm that @jortkoopmans  zip file broke after 11 restarts on my workstation which runs

```
▶09 ~/Downloads/tr 
:56 tf@whale$ uname -a
Linux transwhale 5.0.0-29-generic #31-Ubuntu SMP Thu Sep 12 13:05:32 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
▶09 ~/Downloads/tr 
:56 tf@whale$ docker version
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:51 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:15 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Comment by jortkoopmans** [comment]
Since it has been a while, I re-ran the test on traefik v2.1.4.
To my pleasant surprise, it seems this issue has been resolved by other code refactoring.

I've done a superficial analysis and I believe it is related to the rework done here; https://github.com/containous/traefik/pull/5815
Effectively it is splitting the 'Server' into different parts, one is the ConfigurationWatcher, which is handling the changes in the configuration of container routing. Note that the 'return' line is still there (as mentioned by @gregoryelmajjouti ), but the race condition is not happening anymore in my tests. Presumably that is associated to the splitting of 'Server' / refactoring.

Therefore the problem is resolved for me, I encourage other participants of this thread to re-test their use-cases.

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.

I have to close this one because of its lack of activity :disappointed:

Feel free to re-open it or join our [Community Forum](https://community.containo.us/).

**Comment by luisgcu** [comment]
> I also encountered the : `Unable to find the IP address for the container ... : the server is ignored.` error message.
> 
> After some troubleshooting I realized that my container was misbehaving and was continuously restarting. As such traefik did not have the time to set up a backend for the container.

I got exactly the same problem, it took quite bit of time to realize that was the cause also that container  Grafana show random 404 connection errors.
